import fs from "node:fs";
import path from "node:path";
import { describe, expect, it, type Mock, vi } from "vitest";
import { FilePathError } from "~/src/errors/FilePathError";
import { findPlopPath } from "./findPlopPath";

vi.mock("node:fs");

describe("findPlopPath", () => {
	const namePlopFile = "plopfile.js";
	const dir = "/custom";
	const defaultDir = "/default";

	it("returns path from dir if file exists there", () => {
		const expectedPath = path.join(dir, namePlopFile);
		(fs.existsSync as Mock).mockImplementation((p) => p === expectedPath);

		const result = findPlopPath({ dir, defaultDir, namePlopFile });
		expect(result).toBe(expectedPath);
	});

	it("returns path from defaultDir if file exists there", () => {
		const expectedPath = path.join(defaultDir, namePlopFile);
		(fs.existsSync as Mock).mockImplementation((p) => p === expectedPath);

		const result = findPlopPath({ dir, defaultDir, namePlopFile });
		expect(result).toBe(expectedPath);
	});

	it("throws FilePathError if file is not found", () => {
		(fs.existsSync as Mock).mockReturnValue(false);

		expect(() => findPlopPath({ dir, defaultDir, namePlopFile })).toThrowError(
			new FilePathError(`The file "${namePlopFile}" does not exist.`),
		);
	});
});
